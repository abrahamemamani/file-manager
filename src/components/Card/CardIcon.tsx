import { FC } from "react";

export interface CardIconProps {
  icon: React.ForwardRefExoticComponent<
    Omit<React.SVGProps<SVGSVGElement>, "ref"> & {
      title?: string | undefined;
      titleId?: string | undefined;
    } & React.RefAttributes<SVGSVGElement>
  >;
}

export const CardIcon: FC<CardIconProps> = ({ icon: Icon }) => (
  <Icon className="h-14 text-main" />
);
