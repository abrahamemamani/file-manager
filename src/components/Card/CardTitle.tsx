import { FC } from "react";

export interface CardTitleProps {
  value: string;
}

export const CardTitle: FC<CardTitleProps> = ({ value }) => (
  <span className="text-white text-center text-sm mt-5">{value}</span>
);
