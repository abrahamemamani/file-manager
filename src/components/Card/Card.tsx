"use client";
import React, { ReactNode } from "react";
import { Card as FCard } from "flowbite-react";
import { CardIcon } from "./CardIcon";
import { CardTitle } from "./CardTitle";

export interface CardProps<T> {
  data: T;
  className?: string;
  children: (data: T) => ReactNode;
  onClick?: () => void;
}

const CardComponent = <T,>({
  data,
  className = "",
  children,
  onClick = () => console.log(""),
}: CardProps<T>) => (
  <FCard
    onClick={onClick}
    className={`${className} aspect-square cursor-pointer`}
  >
    <div className="flex flex-col items-center">{children(data)}</div>
  </FCard>
);

export const Card = Object.assign(CardComponent, {
  Icon: CardIcon,
  Title: CardTitle,
});
