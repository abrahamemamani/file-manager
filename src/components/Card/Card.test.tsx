import { afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import { Card, CardProps } from "./Card";

interface ICard {
  name: string;
}

const data: ICard = {
  name: "Card Name",
};

const defaultProps: CardProps<ICard> = {
  data,
  children: () => <></>,
  onClick: () => console.log("onClick"),
};

describe("[components/Card]", () => {
  afterEach(() => {
    cleanup();
  });
  describe("when Card is rendered", () => {
    it("should renderer", () => {
      const { asFragment } = render(<Card {...defaultProps} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("when click on the component", () => {
    it("should call onClick function", () => {
      const props = {
        ...defaultProps,
        onClick: vi.fn(),
      };
      render(<Card {...props} />);
      fireEvent.click(screen.getByTestId("flowbite-card"));
      expect(props.onClick).toHaveBeenCalled();
    });
  });
});
