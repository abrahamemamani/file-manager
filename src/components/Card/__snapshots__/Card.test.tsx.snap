// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[components/Card] > when Card is rendered > should renderer 1`] = `
<DocumentFragment>
  <div
    class="flex rounded-lg border border-gray-200 bg-white shadow-md dark:border-gray-700 dark:bg-gray-800 flex-col  aspect-square cursor-pointer"
    data-testid="flowbite-card"
  >
    <div
      class="flex h-full flex-col justify-center gap-4 p-6"
    >
      <div
        class="flex flex-col items-center"
      />
    </div>
  </div>
</DocumentFragment>
`;
